File: elixir/secret-handshake/secret_handshake.exs
defmodule SecretHandshake do
  use Bitwise, only_operators: true
  @commands {"wink", "double blink", "close your eyes", "jump"}

  @doc """
  Determine the actions of a secret handshake based on the binary
  representation of the given `code`.

  If the following bits are set, include the corresponding action in your list
  of commands, in order from lowest to highest.

  1 = wink
  10 = double blink
  100 = close your eyes
  1000 = jump

  10000 = Reverse the order of the operations in the secret handshake
  """
  @spec commands(code :: integer) :: list(String.t())
  def commands(code) do
    code
      |> Integer.digits(2)
      |> Enum.reverse
      |> Enum.with_index
      |> Enum.reduce([], fn({el, idx}, acc)-> add(idx, el, acc) end)
      |> Enum.reverse
  end

  @spec add(idx :: integer, bit :: integer, accumulator :: list(String.t())) :: list(String.t())
  defp add(idx, 1, acc) when idx > 3 do
    Enum.reverse(acc)
  end

  @spec add(idx :: integer, bit :: integer, accumulator :: list(String.t())) :: list(String.t())
  defp add(_, 0, acc) do
    acc
  end

  @spec add(idx :: integer, bit :: integer, accumulator :: list(String.t())) :: list(String.t())
  defp add(idx, 1, acc) do
   [elem(@commands, idx) | acc]
  end
end

File: go/secret-handshake/secret_handshake.go
package secret

const testVersion = 1

var words = []string{"wink", "double blink", "close your eyes", "jump"}

func Handshake(in uint) []string {
	stack := ToBinaryStack(in)
	reverse := false
	var outStack []string
	for idx, flag := range stack {
		switch {
		case idx > 4:
			continue
		case idx == 4:
			reverse = true
		default:
			// idx < 4
			if flag == 1 {
				outStack = append(outStack, words[idx])
			}
		}
	}
	outStack = CondReverse(outStack, reverse)
	return outStack
}

func Reverse(in []string) []string {
	var out = make([]string, len(in))
	for idx, el := range in {
		out[len(in)-idx-1] = el
	}
	return out
}

func CondReverse(arr []string, reverse bool) []string {
	if reverse {
		return Reverse(arr)
	} else {
		return arr
	}
}

func ToBinaryStack(in uint) (stack []int) {
	n := int(in)
	for n > 0 {
		stack = append(stack, n%2)
		n = n / 2
	}
	return
}
